src/ActionKit/CsrfTokenProvider.php

```diff
@@ -0,0 +1,65 @@
+<?php
+namespace ActionKit;
+use ActionKit\CsrfToken;
+use Exception;
+
+class CsrfTokenProvider {
+
+    public $tokenSessionId = '_csrf_token';
+    public $timeout = 300;
+
+    public function __construct($tokenSessionId = '_csrf_token', $timeout=300){
+        $this->tokenSessionId = $tokenSessionId;
+        $this->timeout = $timeout;
+    }
+
+    public function generateToken() {
+        $token = new CsrfToken($this->tokenSessionId, $this->timeout);
+        $token->time = time();
+        $token->salt = $this->randomString(32);
+        $token->sessid = session_id();
+        $token->ip = $_SERVER['REMOTE_ADDR'];
+
+        $_SESSION[$token->tokenSessionId] = serialize($token);
+
+        $hash = $this->calculateHash($token);
+        return base64_encode($hash);
+    }
+
+    public function checkToken($hashCsrfToken) {
+        $token = $this->getToken();
+        if ($token != null) {
+            if (!$token->checkExpiry($_SERVER['REQUEST_TIME'])) {
+                return false;
+            }
+            $tokenHash = base64_decode($hashCsrfToken);
+            $generatedHash = $this->calculateHash($token);
+            if ($tokenHash and $generatedHash) {
+                return $tokenHash == $generatedHash;
+            }
+        }
+        return false;
+    }
+
+    protected function randomString($len = 10) {
+        $rString = '';
+        $chars = 'ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz23456789';
+        $charsTotal  = strlen($chars);
+        for ($i = 0; $i < $len; $i++) {
+            $rInt = (integer) mt_rand(0, $charsTotal);
+            $rString .= substr($chars, $rInt, 1);
+        }
+        return $rString;
+    }
+    
+    protected function calculateHash($token) {
+        return sha1($_SESSION[$token->tokenSessionId]);
+    }
+    
+    protected function getToken() {
```

Comment:
Since we may store the token with different key in the session array, it's not session id anymore. (I know the original implementation depends on the session id)  :)

```php
static public function loadTokenWithSessionKey($key = '_csrf_token') {
    if (isset($_SESSION[$key])) {
       // return the unserialized Token object
    }
    return false;
}
```

The use case:

```php
$token = CsrfTokenProvider::loadTokenWithSessionKey('_csrf_token');
```